server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/TabletTransactionLog.java

```diff
@@ -0,0 +1,388 @@
+      if (!transactions.isEmpty()) {
+        builder.append("\nFinal files: ").append(new TreeSet<>(Arrays.asList(finalFiles)));
+      }
+
+      return builder.toString();
+    }
+  }
+
+  /**
+   * The max size of the log as derived from the configuration
+   */
+  private static class MaxLogSize {
+    private final int maxSize;
+
+    public MaxLogSize(AccumuloConfiguration config) {
+      maxSize = config.getCount(Property.TABLE_OPERATION_LOG_MAX_SIZE);
+    }
+
+    public int getMaxSize() {
+      return maxSize;
+    }
+  }
+
+  /**
+   * A simple implementation of a ring buffer. Note that this is not thread safe when it comes to
+   * modifications. However the read methods can be done concurrently with a write method.
+   */
+  public static class Ring<T> {
+    private final Object[] ring;
+    private volatile int first;
+    private volatile int last;
+    private static final int overrunThreshold = (Integer.MAX_VALUE / 2);
+
+    public Ring(int size) {
+      ring = new Object[size];
+      first = 0;
+      last = -1;
+    }
+
+    @SuppressWarnings("unchecked")
+    public T add(T object) {
+      Object removed = null;
+      if (ring.length > 0) {
+        if (size() == ring.length) {
+          removed = ring[first++ % ring.length];
+        }
+        int newEnd = last + 1;
+        ring[newEnd % ring.length] = object;
+        last = newEnd;
+      } else {
+        removed = object;
+      }
+      avoidOverrun();
+      return (T) removed;
+    }
+
+    public int size() {
+      return last - first + 1;
+    }
+
+    public int capacity() {
+      return ring.length;
+    }
+
+    public boolean isEmpty() {
+      return last < first;
+    }
+
+    public void avoidOverrun() {
+      if (last > overrunThreshold) {
+        int max = Math.max(first, last);
+        first -= max;
+        last -= max;
+      } else if (first > overrunThreshold) {
+        int max = Math.max(first, last);
+        last -= max;
+        first -= max;
+      }
+    }
+
+    public void clear() {
+      last = first - 1;
+    }
+
+    @SuppressWarnings("unchecked")
+    public List<T> toList() {
+      Object[] data = null;
+      // consistency is defined as the last position and the first position being
+      // the same before and after pulling the list.
+      boolean consistent = false;
+      do {
+        long updateCount = getUpdateCount();
+        int lastPos = last;
+        int firstPos = first;
+        // if either is over the threshold, then we must be in the middle of a modification
+        // but before the overrun threshold is called. try again.
+        if (lastPos <= overrunThreshold && firstPos <= overrunThreshold) {
+          data = new Object[lastPos - firstPos + 1];
+          int index = 0;
+          for (int i = firstPos; i <= lastPos; i++) {
+            data[index++] = ring[i % ring.length];
+          }
+          consistent = (updateCount == getUpdateCount());
+        }
+      } while (!consistent);
```

Comment:
If its not expected that multiple threads will call this code, then detecting unexpected concurrency and throwing an exception could simplify the code a bit.   Also good if the code does not try to handle situations that are not expected to happen.  Took a swag at implementing this as an example of what I was thinking, but not sure if its correct.


```suggestion

        long updateCount = getUpdateCount();
        int lastPos = last;
        int firstPos = first;
        // if either is over the threshold, then we must be in the middle of a modification
        // but before the overrun threshold is called. try again.
        if (lastPos <= overrunThreshold && firstPos <= overrunThreshold) {
          data = new Object[lastPos - firstPos + 1];
          int index = 0;
          for (int i = firstPos; i <= lastPos; i++) {
            data[index++] = ring[i % ring.length];
          }
          consistent = (updateCount == getUpdateCount());
        }
     if(!consistent){
         throw new ConcurrentModificationException();
     }
```

Reply:
Well then I would simply need to put the loop in the places that I call toList() which is in multiple places.  Would you like me to rename this method or create a separate method that does the looping until consistent?

Reply:
I am thinking if an inconsistency is detected that it is an indication of a bug in the code and that an attempt should not be made to handle that, so in that case just let the exception fly.  If that is not the case, then what the expected situation that this code is handling?

Reply:
@ivakegg unresolving this issue because the question I asked was not answer.  I can not review this code if I do not understand the reason some of the code exists.

Reply:
So sorry.  I will try not to do that.  WRT to your question, I wrote this method to be able to run concurrently with the modifications.  It is perfectly reasonable that the metadata consistency checker thread will run the check at the same sime that modifications are being made.  Hence I need to write this method to allow for that.  You can see the test that I added to ensure that this method can be run at the same time as the writers.

Reply:
This question does suggest I need more comments.... I will work on that.

Reply:
Documentation added.  Also added synchronization on the write methods to make this more obvious and safe.

Reply:
I made a comment about a possible negative array size exception that I think could happen.  I also made another comment with a test I added that is failing and I am not sure why.

Adding the synchronization around adding to the ring buffer seems good.  Since reading the list is much a less frequent operation I would add synchronization around it to simplify analysis of this code.  The code seem correct to me in face of concurrency for constructing the list, however the little test I wrote is failing and I am not sure why.  Maybe the test itself has  a flaw in it.

Reply:
Checking out the failing test.  I think I have resolved the questions in this thread.  Mark as resolved if you agree.

Reply:
I see why the test fails.  The toList will return the list of items as long as the update count is the same before and after.  It is possible for the addTransaction to have removed the last item (increased first), and the list is gathered before the item is added to the end.  The question is whether that is ok.  I think it is.

Reply:
> The question is whether that is ok. I think it is.

There is one confusing aspect to that behavior.  If I create a ring buffer to keep N elements and later ask it for a copy of its contents and it gives me N-1 one elements I may think that it has given me everything it ever saw because it had never filled up to N.  With this behavior getting less than N elements is not an indication that everything is being seen.

Reply:
However it was in the process of removing an item.  Think of it this way.  If I split the remove and the add into separate operations, then the toList() would simply be returning what was the state of the ring after I called remove but before I called add.

Reply:
So in the end, is this updateCounter and consistency checking worth avoiding synchronizing the TabletTransactionLog.getTransactions() method ?

Reply:
I am now leaning that way.  The complexity is not worth it.  As long as I move all of the string buffer generation in the dumpLog outside of the synchronized then the impact would be minimal.